### ISSUE TITLE
Revisit get_spent

### ISSUE THREAD

**Issue by r-marques** [bigchaindb/bigchaindb#1380]
The get_spent logic needs to be revisited. As is there is a possibility for nodes to create blocks containing double spends. The block will ultimately be marked as invalid but this leads to unnecessary create block, mark invalid, rewrite transaction to the backlog loops.

There two approaches for get_spent:
- Return only if the spent is in a `VALID` transaction. This would simplify the logic since get_spent would always return at most on result.
- Return if the output was spend in transactions in either `VALID`, `UNDECIDED` or `BACKLOG`. This is the current behaviour. The logic is a bit more complicated but the advantage is that we are able to prevent double spends from ever being included in the blocks. The problem is that it leads to non deterministic behaviours, mainly because of these two lines of code https://github.com/bigchaindb/bigchaindb/blob/v0.9.5/bigchaindb/core.py#L369 and https://github.com/bigchaindb/bigchaindb/blob/v0.9.5/bigchaindb/models.py#L81. This means that whatever is returned by `transactions[0]` will determine which transaction will be considered valid and which transaction(s) will be considered invalid.

A good way to test this behaviour is to check what happens when there are two different transactions in the backlog that spend the same output using the [stepping pipeline](https://github.com/bigchaindb/bigchaindb/blob/bc5217b9f31f5d1601e5f5c00524c4a438ef351d/tests/pipelines/stepping.py) and try to emulate both a sequential and concurrent behaviour.

**Comment by ttmc** [comment]
We might even consider a third approach for get_spent:

* Return if the output was spent in a `VALID` transaction _in the past_, i.e. in an earlier block. Don't worry about transactions in future blocks. (This requires a block order shared by all nodes, which is something planned in the medium-term roadmap.)

**Comment by r-marques** [comment]
Yes that is one approach. The problem is that with that approach if we have multiple transactions in the backlog that spend the same output they will all be included in a block because that output was not spent _in the past_.

**Comment by ttmc** [comment]
Yes, good point... I should revise the approach a bit:

Let's assume that there's an agreed-upon block order shared by all nodes, so that before a node can determine the validity of block N, it must first know (or be able to look up) the validity of blocks 0, 1, 2, ..., and (N-1).

* When checking the validity of block N: for each transaction in block N, check if it spends an output that was already spent by:
    * a transaction in blocks 0...(N-1), or
    * a different transaction in block N.

This approach can still be summarized as "When checking for double spends, don't worry about transactions in future blocks."
